Add plugin: Notemd

Jacobinwwey opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin:https://github.com/Jacobinwwey/obsidian-NotEMD

## Release Checklist
- [] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

## Description
A Easy way to create your own Knowledge-base!

Notemd enhances your Obsidian workflow by integrating with various Large Language Models (LLMs) to process your notes, automatically generate wiki-links for key concepts, create corresponding concept notes, perform web research, and more.
## Features

### AI-Powered Document Processing
- **Multi-LLM Support**: Connect to various cloud and local LLM providers (see [Supported LLM Providers](#supported-llm-providers)).
- **Smart Chunking**: Automatically splits large documents into manageable chunks based on word count for processing.
- **Content Preservation**: Aims to maintain original formatting while adding structure and links.
- **Progress Tracking**: Real-time updates via the Notemd Sidebar or a progress modal.
- **Cancellable Operations**: Cancel any processing task (single or batch) initiated from the sidebar via its dedicated cancel button. Command palette operations use a modal which can also be cancelled.
- **Multi-Model Configuration**: Use different LLM providers *and* specific models for different tasks (Add Links, Research, Generate Title) or use a single provider for all.
- **Stable API Calls (Retry Logic)**: Optionally enable automatic retries for failed LLM API calls with configurable interval and attempt limits.

### Knowledge Graph Enhancement
- **Automatic Wiki-Linking**: Identifies and adds `[[wiki-links]]` to core concepts within your processed notes based on LLM output.
- **Concept Note Creation (Optional & Customizable)**: Automatically creates new notes for discovered concepts in a specified vault folder.
- **Customizable Output Paths**: Configure separate relative paths within your vault for saving processed files and newly created concept notes.
- **Customizable Output Filenames (Add Links)**: Optionally **overwrite the original file** or use a custom suffix/replacement string instead of the default `_processed.md` when processing files for links.
- **Link Integrity Maintenance**: Basic handling for updating links when notes are renamed or deleted within the vault.

### Web Research & Content Generation
- **Web Research & Summarization**:
    - Perform web searches using Tavily (requires API key) or DuckDuckGo (experimental).
    - Summarize search results using the configured LLM.
    - Append summaries to the current note.
    - Configurable token limit for research content sent to the LLM.
- **Content Generation from Title**:
    - Use the note title to generate initial content via LLM, replacing existing content.
    - **Optional Research**: Configure whether to perform web research (using the selected provider) to provide context for generation.
- **Batch Content Generation from Titles**: Generate content for all notes within a selected folder based on their titles (respects the optional research setting). Successfully processed files are moved to a **configurable "complete" subfolder** (e.g., `[foldername]_complete` or a custom name) to avoid reprocessing.

### Utility Features
- **Duplicate Detection**: Basic check for duplicate words within the currently processed file's content (results logged to console).
- **Check and Remove Duplicate Concept Notes**: Identifies potential duplicate notes within the configured **Concept Note Folder** based on exact name matches, plurals, normalization, and single-word containment compared to notes outside the folder. The scope of the comparison (which notes outside the concept folder are checked) can be configured to the **entire vault**, **specific included folders**, or **all folders excluding specific ones**. Presents a detailed list with reasons and conflicting files, then prompts for confirmation before moving identified duplicates to system trash. Shows progress during deletion.
- **Batch Mermaid Fix**: Applies Mermaid and LaTeX syntax corrections (`refineMermaidBlocks` and `cleanupLatexDelimiters`) to all Markdown files within a user-selected folder.
- **LLM Connection Test**: Verify API settings for the active provider.

joethei commented:
[name: 'Process Current File (Add Links)',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L59), [name: 'Process Current File (Add Links)',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L59), [name: 'Check for Duplicates in Current File',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L76), [name: 'Test LLM Connection',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L106), [name: 'Generate Content from Note Title',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L114), [name: 'Research and Summarize Topic',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L127), [name: 'Batch Generate Content from Titles',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L136), [name: 'Check and Remove Duplicate Concept Notes',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L144), [name: 'Batch Fix Mermaid Syntax',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L152), [contentEl.createEl('h3', { text: 'Notemd Processing' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/ProgressModal.ts#L26), [container.createEl("h4", { text: "Original Processing" });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L196), [this.processCurrentButton = originalButtonGroup.createEl('button', { text: 'Process File (Add Links)', cls: 'mod-cta' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L199), [this.processFolderButton = originalButtonGroup.createEl('button', { text: 'Process Folder (Add Links)' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L211), [container.createEl('h4', { text: "New Features" });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L223), [this.batchGenerateTitleButton = newFeatureButtonGroup.createEl('button', { text: 'Batch Generate from Titles' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L289), [this.batchMermaidFixButton = utilityButtonGroup.createEl('button', { text: 'Batch Mermaid Fix' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L304), [this.checkDuplicatesButton = utilityButtonGroup.createEl('button', { text: 'Check Duplicates (Current File)' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L316), [this.testConnectionButton = utilityButtonGroup.createEl('button', { text: 'Test LLM Connection' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L347), [this.checkRemoveDuplicatesButton = utilityButtonGroup.createEl('button', { text: 'Check & Remove Duplicates' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L354), [this.cancelButton = progressArea.createEl('button', { text: 'Cancel Processing', cls: 'notemd-cancel-button' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L373), [logHeader.createEl('h5', { text: 'Log Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L378), [const copyLogButton = logHeader.createEl('button', { text: 'Copy Log', cls: 'notemd-copy-log-button' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L379), [.setName('Manage Provider Configurations')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L115), ['Export Providers'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L118), ['Import Providers'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L120), [.setName('Active Provider')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L123), [.setName('API Key')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L144), [.setName('API Version')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L179), [.setName(`Test ${activeProvider.name} Connection`)](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L188), [.setName("Remove Code Fences on 'Add Links'")](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L284), [containerEl.createEl('h4', { text: 'Concept Note Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L293), [containerEl.createEl('h4', { text: 'Concept Log File Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L299), [('Customize Concept Note Path')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L294), ['Concept Note Folder Path'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L296), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), ['Customize Log File Name'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L310), ['Concept Log File Name'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L314), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), ['Enable Research in "Generate from Title"')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L319), ["Use Custom Output Folder for 'Generate from Title'"](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L320), ["Custom Output Folder Name"](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L322), [containerEl.createEl('h4', { text: 'Web Research Provider' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L325), ['Search Provider'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L326), ['Tavily API Key'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L328), ['Tavily Max Results'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L329), ['Tavily Search Depth'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L330), ['DuckDuckGo Max Results'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L332), ['DuckDuckGo Content Fetch Timeout (seconds)'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L333), [('Max Research Content Tokens'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L335), [containerEl.createEl('h4', { text: 'Processing Parameters' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L337), ['Chunk Word Count'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L338), ['Enable Duplicate Detection'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L339), ['Max Tokens'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L340), [containerEl.createEl('h4', { text: 'Duplicate Check Scope' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L343), [.setName('Duplicate Check Scope Mode')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L346), ['Entire Vault](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L349), ['Include Specific Folders Only](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L350), ['Exclude Specific Folders](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L351), ['Concept Folder Only](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L352), [modal.titleEl.setText('Confirm Duplicate Deletion');](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/modals.ts#L12)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[callback: async () => {](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L77), [callback: async () => {](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L115)
These should be using [conditional commands](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)

[name: 'Batch Fix Mermaid Syntax',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L152)
Why is this command a part of your plugin?
It makes no sense to me.

[containerEl.createEl('h2', { text: 'Notemd Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L109)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[containerEl.createEl('h3', { text: 'LLM Provider Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L112), [containerEl.createEl('h3', { text: 'Multi-Model Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L214), [containerEl.createEl('h3', { text: 'Stable API Call Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L260), [containerEl.createEl('h4', { text: 'Concept Note Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L293), [containerEl.createEl('h4', { text: 'Concept Log File Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L299), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), [containerEl.createEl('h4', { text: 'Web Research Provider' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L325), [containerEl.createEl('h4', { text: 'Processing Parameters' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L337), [containerEl.createEl('h4', { text: 'Duplicate Check Scope' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L343)
For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

[containerEl.createEl('h3', { text: 'LLM Provider Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L112), [containerEl.createEl('h3', { text: 'Multi-Model Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L214)
[Avoid "configuration" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[containerEl.createEl('h3', { text: 'Stable API Call Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L260)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

Jacobinwwey commented:
1. Since note generation requires a large language model to generate Mermaid diagrams, the model often encounters formatting errors in complex Mermaid code, such as missing closing brackets (‘;] or ]’) in Mermaid code blocks. This command helps correct LLM-generated or user-written Mermaid/LaTeX syntax errors in batch, ensuring diagrams and equations render correctly, aligning with the plugin's theme of note enhancement and quality maintenance. ‘Batch Fix Mermaid Syntax’ will perform preliminary format corrections on easily identifiable and manageable parts in such cases, facilitating batch script processing for users.
2. The requested code adjustments have been made:

- In `ui/NotemdSettingTab.ts`:

- Removed the top-level `<h2>` ‘Notemd Settings’.
- Converted `<h3>` and `<h4>` subheadings to use `new Setting().setHeading()`.
- Updated setting names and button texts to use sentence case (e.g., ‘Active provider,’ ‘Export providers’).
- Refined heading names to be more concise (e.g., ‘LLM Provider Configuration’ became ‘LLM providers’).

- In `main.ts`:
- Updated all command names (e.g., `addCommand({ name: ‘...’ })`) to use sentence case (e.g., ‘Process current file (add links),’ ‘Batch fix Mermaid syntax’).

- In `ui/modals.ts`:

- Updated modal title `Confirm Duplicate Deletion` to `Confirm duplicate deletion`.
- Updated button text `Delete ${reportList.length} Files` to `Delete ${reportList.length} files`.

- In `ui/NotemdSidebarView.ts`:
- Updated button texts and section headings (e.g., `h4`) to use sentence case (e.g., ‘Process file (add links),’ ‘Original processing,’ ‘Log output’).

> [name: 'Process Current File (Add Links)',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L59), [name: 'Process Current File (Add Links)',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L59), [name: 'Check for Duplicates in Current File',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L76), [name: 'Test LLM Connection',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L106), [name: 'Generate Content from Note Title',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L114), [name: 'Research and Summarize Topic',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L127), [name: 'Batch Generate Content from Titles',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L136), [name: 'Check and Remove Duplicate Concept Notes',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L144), [name: 'Batch Fix Mermaid Syntax',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L152), [contentEl.createEl('h3', { text: 'Notemd Processing' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/ProgressModal.ts#L26), [container.createEl("h4", { text: "Original Processing" });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L196), [this.processCurrentButton = originalButtonGroup.createEl('button', { text: 'Process File (Add Links)', cls: 'mod-cta' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L199), [this.processFolderButton = originalButtonGroup.createEl('button', { text: 'Process Folder (Add Links)' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L211), [container.createEl('h4', { text: "New Features" });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L223), [this.batchGenerateTitleButton = newFeatureButtonGroup.createEl('button', { text: 'Batch Generate from Titles' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L289), [this.batchMermaidFixButton = utilityButtonGroup.createEl('button', { text: 'Batch Mermaid Fix' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L304), [this.checkDuplicatesButton = utilityButtonGroup.createEl('button', { text: 'Check Duplicates (Current File)' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L316), [this.testConnectionButton = utilityButtonGroup.createEl('button', { text: 'Test LLM Connection' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L347), [this.checkRemoveDuplicatesButton = utilityButtonGroup.createEl('button', { text: 'Check & Remove Duplicates' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L354), [this.cancelButton = progressArea.createEl('button', { text: 'Cancel Processing', cls: 'notemd-cancel-button' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L373), [logHeader.createEl('h5', { text: 'Log Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L378), [const copyLogButton = logHeader.createEl('button', { text: 'Copy Log', cls: 'notemd-copy-log-button' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSidebarView.ts#L379), [.setName('Manage Provider Configurations')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L115), ['Export Providers'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L118), ['Import Providers'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L120), [.setName('Active Provider')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L123), [.setName('API Key')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L144), [.setName('API Version')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L179), [.setName(`Test ${activeProvider.name} Connection`)](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L188), [.setName("Remove Code Fences on 'Add Links'")](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L284), [containerEl.createEl('h4', { text: 'Concept Note Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L293), [containerEl.createEl('h4', { text: 'Concept Log File Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L299), [('Customize Concept Note Path')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L294), ['Concept Note Folder Path'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L296), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), ['Customize Log File Name'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L310), ['Concept Log File Name'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L314), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), ['Enable Research in "Generate from Title"')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L319), ["Use Custom Output Folder for 'Generate from Title'"](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L320), ["Custom Output Folder Name"](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L322), [containerEl.createEl('h4', { text: 'Web Research Provider' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L325), ['Search Provider'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L326), ['Tavily API Key'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L328), ['Tavily Max Results'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L329), ['Tavily Search Depth'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L330), ['DuckDuckGo Max Results'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L332), ['DuckDuckGo Content Fetch Timeout (seconds)'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L333), [('Max Research Content Tokens'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L335), [containerEl.createEl('h4', { text: 'Processing Parameters' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L337), ['Chunk Word Count'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L338), ['Enable Duplicate Detection'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L339), ['Max Tokens'](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L340), [containerEl.createEl('h4', { text: 'Duplicate Check Scope' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L343), [.setName('Duplicate Check Scope Mode')](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L346), ['Entire Vault](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L349), ['Include Specific Folders Only](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L350), ['Exclude Specific Folders](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L351), ['Concept Folder Only](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L352), [modal.titleEl.setText('Confirm Duplicate Deletion');](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/modals.ts#L12) [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)
> 
> [callback: async () => {](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L77), [callback: async () => {](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L115) These should be using [conditional commands](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)
> 
> [name: 'Batch Fix Mermaid Syntax',](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/main.ts#L152) Why is this command a part of your plugin? It makes no sense to me.
> 
> [containerEl.createEl('h2', { text: 'Notemd Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L109) [Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)
> 
> [containerEl.createEl('h3', { text: 'LLM Provider Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L112), [containerEl.createEl('h3', { text: 'Multi-Model Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L214), [containerEl.createEl('h3', { text: 'Stable API Call Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L260), [containerEl.createEl('h4', { text: 'Concept Note Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L293), [containerEl.createEl('h4', { text: 'Concept Log File Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L299), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), [containerEl.createEl('h4', { text: 'Content Generation & Output' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L318), [containerEl.createEl('h4', { text: 'Web Research Provider' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L325), [containerEl.createEl('h4', { text: 'Processing Parameters' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L337), [containerEl.createEl('h4', { text: 'Duplicate Check Scope' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L343) For section headings in settings use:
> 
> ```ts
> new Setting(containerEl).setName('name here').setHeading();
> ```
> 
> [containerEl.createEl('h3', { text: 'LLM Provider Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L112), [containerEl.createEl('h3', { text: 'Multi-Model Configuration' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L214) [Avoid "configuration" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)
> 
> [containerEl.createEl('h3', { text: 'Stable API Call Settings' });](https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/NotemdSettingTab.ts#L260) [Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

Jacobinwwey commented:
I have completed the changes you requested and look forward to hearing from you soon. @joethei

joethei commented:
You missed this one: https://github.com/Jacobinwwey/obsidian-NotEMD/blob/d7301b4fdf17a082acc77f340a24d353f91693bc/ui/ProgressModal.ts#L26

Jacobinwwey commented:
Sorry, I have re-reviewed the entire code base as requested and have now made the necessary changes. @joethei
